### ISSUE TITLE
GMailOAuth2 in VB.NET

### ISSUE THREAD

**Issue by gt2emme** [jstedfast/MailKit#1370]
I'm trying to convert the code in [https://github.com/jstedfast/MailKit/blob/master/GMailOAuth2.md] from C# to VB.NET but, apart from not being sure that I have done it correctly, I have an issue with this bit in particular:

var credential = await authCode.AuthorizeAsync (GMailAccount, CancellationToken.None);

VB doesn't like the "await" and most of all it doesn't see Token as part of credential. 
In other words, credential is not working properly therefore I can't use the code.

Does anybody know if there is a VB version of this or can help me in any way?

Thanks.

**Comment by jstedfast** [comment]
`await` in VB is just `Await`, but you also need to make sure that the method that you are copying this code into is marked as `Async`.

**Comment by gt2emme** [comment]
Thank you Jeffrey.

 

I do not control the authCode.AuthorizeAsync method though.

 

I think the main issue is that in VB authCode.AuthorizeAsync (which is part of Google Apis AuthorizationCodeInstalledApp) doesn’t return the proper object.

 

Would you know of a VB adaptation of some sort of this code?

 

Thanks.

 

Gianpiero

 

Da: Jeffrey Stedfast ***@***.***> 
Inviato: mercoledì 27 aprile 2022 19:04
A: jstedfast/MailKit ***@***.***>
Cc: gt2emme ***@***.***>; Author ***@***.***>
Oggetto: Re: [jstedfast/MailKit] GMailOAuth2 in VB.NET (Issue #1370)

 

await in VB is just Await, but you also need to make sure that the method that you are copying this code into is marked as Async.

**Comment by jstedfast** [comment]
You don't need to control the authCode.AuthorizeAsync method, you just need to control the method that calls it (which is your code).

Either that or just call `authCode.AuthorizeAsync().GetAwaiter().GetResult()`.

**Comment by gt2emme** [comment]
Great! Many thanks.

 

Have a good day.

 

Gianpiero

 

Da: Jeffrey Stedfast ***@***.***> 
Inviato: mercoledì 27 aprile 2022 19:47
A: jstedfast/MailKit ***@***.***>
Cc: gt2emme ***@***.***>; Author ***@***.***>
Oggetto: Re: [jstedfast/MailKit] GMailOAuth2 in VB.NET (Issue #1370)

 

You don't need to control the authCode.AuthorizeAsync method, you just need to control the method that calls it (which is your code).

Either that or just call authCode.AuthorizeAsync().GetAwaiter().GetResult().

**Comment by jstedfast** [comment]
Just replace ImapClient with SmtpClient. That's all you need to do.

**Comment by gt2emme** [comment]
Thank you for your reply Jeffrey.

 

If you mean when I create the object like in 

 

SMTP = New MailKit.Net.Smtp.SmtpClient()

 

I’ve done it already.

 

The authentication works

 

                    SMTP.AuthenticateAsync(oauth2)

                    SMTP.ConnectAsync("imap.gmail.com", 993, SecureSocketOptions.SslOnConnect)

 

But when I send the email I get an error message saying that I should verify account and password.

 

Gianpiero

 

Da: Jeffrey Stedfast ***@***.***> 
Inviato: lunedì 2 maggio 2022 22:24
A: jstedfast/MailKit ***@***.***>
Cc: gt2emme ***@***.***>; Author ***@***.***>
Oggetto: Re: [jstedfast/MailKit] GMailOAuth2 in VB.NET (Issue #1370)

 

Just replace ImapClient with SmtpClient. That's all you need to do.

**Comment by ekalchev** [comment]
You should call Connect first then Authenticate. If you still get the error you really "should verify account and password"

**Comment by jstedfast** [comment]
There are a few things wrong with your code snippet. As @ekalchev points out, you need to Connect first, and then Authenticate (you had it backwords).

Second, you need to connect to the correct host/port.

```vb.net
SMTP = New MailKit.Net.Smtp.SmtpClient()
SMTP.ConnectAsync("smtp.gmail.com", 465, SecureSocketOptions.SslOnConnect)
SMTP.AuthenticateAsync(oauth2)
```

**Comment by gt2emme** [comment]
It worked, thanks a lot.

 

Kind regards

 

Gianpiero

 

Da: Jeffrey Stedfast ***@***.***> 
Inviato: giovedì 5 maggio 2022 15:35
A: jstedfast/MailKit ***@***.***>
Cc: gt2emme ***@***.***>; Author ***@***.***>
Oggetto: Re: [jstedfast/MailKit] GMailOAuth2 in VB.NET (Issue #1370)

 

You need to Await ConnectAsync() and Await AuthenticateAsync() - or just don't use the Async versions of the methods.
